reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,129 @@
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(job + dispatcher)
+
+    override fun scheduleDirect(block: Runnable): Disposable {
+        if (scope.isActive) {
+            return scope.launch {
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (delay <= 0) {
+            return scheduleDirect(block)
+        }
+
+        if (scope.isActive) {
+            return scope.launch {
+                delay(unit.toMillis(delay))
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    private fun Runnable.runWithRx() = RxJavaPlugins.onSchedule(this).run()
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, job)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerScope = CoroutineScope(SupervisorJob(parentJob) + dispatcher)
+        private val blockChannel = Channel<Job>(Channel.UNLIMITED)
+        private var queueProcessingJob: Job? = null
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable {
+            startProcessingQueue()
+            if (workerScope.isActive) {
+                val job = workerScope.launch(start = CoroutineStart.LAZY) {
+                    block.run()
+                }
+                blockChannel.offer(job)
+                return job.asDisposable()
+            }
+
+            return Disposables.disposed()
+        }
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            startProcessingQueue()
+            if (delay <= 0) {
+                return schedule(block)
+            }
+            if (workerScope.isActive) {
+                return workerScope.launch {
+                    delay(unit.toMillis(delay))
+                    block.run()
```

Comment:
@zach-klippenstein I'll also make this this on the channel.

As far as cancelling the task, wouldn't the solution I have to schedule without delay work? Right now I put the Job/Disposable itself on the queue. Then I check if the job is active once we dequeue from the channel. Wouldn't that handle what you're referring to? Let me know if I'm misunderstanding.

Reply:
Yea, I think it would work, but creating a new coroutine + job for that isn't necessary (the job isn't needed for structured concurrency, hierarchy, or cancelling a running coroutine – just checking if the task was disposed while waiting in the queue). But I think it should still be correct, I would just add some comments to explain why the launch is lazy and in what cases the job would be cancelled, since it all looks a bit strange at a first glance.